pkg/kubelet/server/server.go

```diff
@@ -369,33 +367,36 @@ func (s *Server) InstallDefaultHandlers() {
 		cadvisormetrics.ProcessMetrics:      struct{}{},
 	}
 
-	// Only add the Accelerator metrics if the feature is inactive
+	// Only add the Accelerator metrics if the feature is inactive.
 	// Note: Accelerator metrics will be removed in the future, hence the feature gate.
 	if !utilfeature.DefaultFeatureGate.Enabled(features.DisableAcceleratorUsageMetrics) {
 		includedMetrics[cadvisormetrics.AcceleratorUsageMetrics] = struct{}{}
 	}
 
+	// cAdvisor metrics are exposed under the secured handler as well.
+	container := metrics.NewContainerCollector(prometheusHostAdapter{s.host}, containerPrometheusLabelsFunc(s.host), includedMetrics, clock.RealClock{})
+	machine := metrics.NewMachineCollector(prometheusHostAdapter{s.host}, includedMetrics)
+
 	cadvisorOpts := cadvisorv2.RequestOptions{
 		IdType:    cadvisorv2.TypeName,
 		Count:     1,
 		Recursive: true,
+		// Set MaxAge to ensure that scrapes coming within 5 seconds does not need to re-update cache (e.g. two-replica Prometheus scrapes).
+		// This allows some savings in kubelet CPU usage. Typical Prometheus configuration scrapes this in longer intervals.
+		MaxAge: func() *time.Duration { maxAge := 5 * time.Second; return &maxAge }(),
 	}
-	r.RawMustRegister(metrics.NewPrometheusCollector(prometheusHostAdapter{s.host}, containerPrometheusLabelsFunc(s.host), includedMetrics, clock.RealClock{}, cadvisorOpts))
-	r.RawMustRegister(metrics.NewPrometheusMachineCollector(prometheusHostAdapter{s.host}, includedMetrics))
-	s.restfulCont.Handle(cadvisorMetricsPath,
-		compbasemetrics.HandlerFor(r, compbasemetrics.HandlerOpts{ErrorHandling: compbasemetrics.ContinueOnError}),
-	)
+	s.restfulCont.Handle(cadvisorMetricsPath, promhttp.HandlerForTransactional(
+		metrics.UpdateOnMaxAgeGatherer(cadvisorOpts, metrics.NewCachedGatherer(container.Collect, machine.Collect)),
+		promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError},
+	))
 
-	s.addMetricsBucketMatcher("metrics/resource")
 	resourceRegistry := compbasemetrics.NewKubeRegistry()
 	resourceRegistry.CustomMustRegister(collectors.NewResourceMetricsCollector(s.resourceAnalyzer))
 	s.restfulCont.Handle(resourceMetricsPath,
 		compbasemetrics.HandlerFor(resourceRegistry, compbasemetrics.HandlerOpts{ErrorHandling: compbasemetrics.ContinueOnError}),
 	)
 
-	// prober metrics are exposed under a different endpoint
-
-	s.addMetricsBucketMatcher("metrics/probes")
```

Comment:
Looks like these unrelated metrics buckets were removed @bwplotka, was that intentional? We still need these for non-cadvisor provided metrics.